File: src/eQuantic.UI.Runtime/src/shared/components/RadioGroup.ts
import { Adjustable, Box, BoxStyle, BuildContext, Column, CornerRadii, Pressable, Row, SizeValue, Sizing, StatelessComponent, Text, VisualNode } from "../runtime-exports";

export class RadioGroup extends StatelessComponent {
    declare options: string[];
    declare selected: number;
    declare onChanged: any;
    declare label: any;
    declare disabled: boolean;
    constructor(options?: any, selected?: any, onChanged: any = null, label: any = null, props?: any) {
        super();
        if (options !== undefined) this.options = options;
        if (selected !== undefined) this.selected = selected;
        if (onChanged !== undefined) this.onChanged = onChanged;
        if (label !== undefined) this.label = label;
        if (this.selected === undefined) this.selected = 0;
        if (this.disabled === undefined) this.disabled = false;
        this.options = options;this.selected = selected;this.onChanged = onChanged;this.label = label;
        if (props && typeof props === 'object') Object.assign(this, props);
    }

    build(context: BuildContext) {
        let theme = context.theme;let primary = theme.colors('primary');let options = new Column(4, 'start', 'stretch', false, null, null, { width: SizeValue.fill });for (let i = 0; i < this.options.length; i++) {let isSelected = i === this.selected;let index = i;let circleContent = isSelected ? new Box(new BoxStyle({ width: Sizing.radioDot(context.density), height: Sizing.radioDot(context.density), background: primary.base, cornerRadius: new CornerRadii(theme.shape('full')) })).centered() : null;let circle = new Box(new BoxStyle({ width: Sizing.selectionBox(context.density), height: Sizing.selectionBox(context.density), cornerRadius: new CornerRadii(theme.shape('full')), borderWidth: 2, borderColor: isSelected ? primary.base : theme.borderStrong }), circleContent);let row = new Row(12, 'start', 'center', false, null, null, { cross: 'center', width: SizeValue.fill, height: 44 });row.add(circle);row.add(new Text(this.options[i], 'bodyM', this.disabled ? theme.textMuted : theme.textPrimary, 1));options.add(new Pressable(row, this.disabled || this.onChanged == null || isSelected ? null : () => this.onChanged(index), { disabled: this.disabled, label: this.options[i], role: 'radio', selected: isSelected }));}let group: VisualNode = options;if (!this.disabled && !(this.onChanged == null) && this.options.length > 0) {let count = this.options.length;group = new Adjustable(options, (direction: number) => this.onChanged((this.selected + direction + count) % count), { role: 'radiogroup', label: this.label });}let column = new Column(4, 'start', 'stretch', false, null, null, { width: SizeValue.fill });let groupLabel: any; if ((groupLabel = this.label) != null) column.add(new Text(groupLabel, 'caption', theme.textMuted, 1));column.add(group);return column;
    }

}

File: src/eQuantic.UI.Components/RadioGroup.cs
using eQuantic.UI.Primitives;

namespace eQuantic.UI.Components;

/// <summary>
/// The design system's RadioGroup (spec B13), CONTROLLED: a group always has exactly one value —
/// selection moves, never clears. Radio = 22dp circle: 2dp BorderStrong; selected = 2dp Primary
/// ring + 10dp Primary dot (two rrect draws). Each row is a full-width target, min 44 tall,
/// gap 12. ONE Tab stop for the whole group — the <see cref="Adjustable"/> wrapper — and the
/// arrows walk the choice with wrap-around, exactly the native <c>&lt;input type=radio&gt;</c>
/// contract; each row is a <c>role="radio"</c> with its checked-ness stated, out of the tab
/// order, still pressable by pointer. v1 fences: the dot scale motion joins the animation
/// system; <c>aria-activedescendant</c> joins the shared id machinery.
/// </summary>
public sealed class RadioGroup : StatelessComponent
{
    public RadioGroup(IReadOnlyList<string> options, int selected, Action<int>? onChanged = null,
        string? label = null)
    {
        Options = options;
        Selected = selected;
        OnChanged = onChanged;
        Label = label;
    }

    public IReadOnlyList<string> Options { get; init; }
    public int Selected { get; init; }
    public Action<int>? OnChanged { get; init; }

    /// <summary>Group label (Caption, TextSecondary) rendered above the options.</summary>
    public string? Label { get; init; }

    public bool Disabled { get; init; }

    public override VisualNode Build(ComponentContext context)
    {
        var theme = context.Theme;
        var primary = theme.Colors(Variant.Primary);

        var options = new Column(gap: Space.S1) { Width = SizeValue.Fill };
        for (var i = 0; i < Options.Count; i++)
        {
            var isSelected = i == Selected;
            var index = i;

            var circleContent = isSelected
                ? new Box(new BoxStyle
                {
                    Width = Sizing.RadioDot(context.Density),
                    Height = Sizing.RadioDot(context.Density),
                    Background = primary.Base,
                    CornerRadius = new CornerRadii(theme.Shape(ShapeScale.Full)),
                }).Centered()
                : null;

            var circle = new Box(new BoxStyle
            {
                Width = Sizing.SelectionBox(context.Density),
                Height = Sizing.SelectionBox(context.Density),
                CornerRadius = new CornerRadii(theme.Shape(ShapeScale.Full)),
                BorderWidth = 2f,
                BorderColor = isSelected ? primary.Base : theme.BorderStrong,
            }, circleContent);

            var row = new Row(gap: Space.S3) { Cross = CrossAlign.Center, Width = SizeValue.Fill, Height = 44 };
            row.Add(circle);
            row.Add(new Text(Options[i], TypeRole.BodyM,
                Disabled ? theme.TextMuted : theme.TextPrimary, maxLines: 1));

            // The selected row presses to nothing, exactly as a native radio does — and each row
            // states what it is: a radio, checked or not, outside the tab order (the group is
            // the Tab stop).
            options.Add(new Pressable(row,
                Disabled || OnChanged is null || isSelected ? null : () => OnChanged(index))
            {
                Disabled = Disabled,
                Label = Options[i],
                Role = PressableRole.Radio,
                Selected = isSelected,
            });
        }

        // The group is ONE Tab stop and the arrows move the choice, wrapping at the ends — the
        // native radio group contract, and the same mechanism the SegmentedControl rides.
        VisualNode group = options;
        if (!Disabled && OnChanged is not null && Options.Count > 0)
        {
            var count = Options.Count;
            group = new Adjustable(options,
                direction => OnChanged((Selected + direction + count) % count))
            {
                Role = AdjustableRole.Radiogroup,
                Label = Label,
            };
        }

        var column = new Column(gap: Space.S1) { Width = SizeValue.Fill };
        if (Label is { } groupLabel)
            column.Add(new Text(groupLabel, TypeRole.Caption, theme.TextMuted, maxLines: 1));
        column.Add(group);
        return column;
    }
}
